### ISSUE TITLE
Rename Operation On Windows

### ISSUE THREAD

**Issue by freshmanken** [xetdata/nfsserve#24]
Hi,

I am trying to implement ReadWrite NFS server on Windows, most of the other operation can be done successfully except for the `rename()` operation. 

So I tried several different behavior:
1. Use `Rename-Item` to rename a file, it actually called multiple interface such as lookup(), create() and remove() to achieve the result.
2. However if I use `Rename-Item` to rename a folder, it always give me access denied, even if I setup all file / folders as 777
3. When I use Windows GUI to rename file / folder through right click, it gives me `invalid device` pop up window

For case 1 is it expected or should the server directly call rename() interface instead? For 2 and 3 could you please confirm rename() should be working on Windows and appreciate if you could provide any insight on what might go wrong? Thanks!

**Comment by ylow** [comment]
Wierd. It should go through the `nfsproc3_rename` method which is implemented. Do you see the debug log message written here? https://github.com/xetdata/nfsserve/blob/2f9c21bbbd19f97f38d39f655aee3877c90d8575/src/nfs_handlers.rs#L1732 

It is possible that there is a bug in the rename implementation. If you can provide some logging information that would help!

**Comment by freshmanken** [comment]
Hi @ylow, here is the log trace for Rename-Item on a file that is successful:

```
    2024-05-07T17:23:19.706463Z  INFO nfsserve::tcp: Accepting connection from 127.0.0.1:922
    2024-05-07T17:23:19.706698Z DEBUG nfsserve::tcp: Accepting socket PollEvented { io: Some(TcpStream { addr: 127.0.0.1:111, peer: 127.0.0.1:922, socket: 728 }) } RPCContext { local_port: 111, client_addr: "127.0.0.1:922", auth: auth_unix { stamp: 0, machinename: [], uid: 0, gid: 0, gids: [] } }
    2024-05-07T17:23:19.708141Z TRACE nfsserve::rpcwire: Reading fragment length:40, last:true
    2024-05-07T17:23:19.708291Z TRACE nfsserve::rpcwire: Finishing Reading fragment length:40, last:true
    2024-05-07T17:23:19.708431Z DEBUG nfsserve::mount_handlers: mountproc3_export(556335)
    2024-05-07T17:23:19.708588Z TRACE nfsserve::rpcwire: Writing fragment length:44
    2024-05-07T17:23:19.708810Z TRACE nfsserve::rpcwire: Reading fragment length:40, last:true
    2024-05-07T17:23:19.708954Z TRACE nfsserve::rpcwire: Finishing Reading fragment length:40, last:true
    2024-05-07T17:23:19.709111Z DEBUG nfsserve::mount_handlers: mountproc3_export(556336)
    2024-05-07T17:23:19.709272Z TRACE nfsserve::rpcwire: Writing fragment length:44
    2024-05-07T17:23:19.709473Z TRACE nfsserve::rpcwire: Reading fragment length:100, last:true
    2024-05-07T17:23:19.709633Z TRACE nfsserve::rpcwire: Finishing Reading fragment length:100, last:true
    2024-05-07T17:23:19.709770Z DEBUG nfsserve::mount_handlers: mountproc3_mnt(556337,"//New folder/New folder")
    [2024-05-07T17:23:19Z INFO  myfs::mainloop_nfs] lookup(1, "New folder")
    [2024-05-07T17:23:19Z INFO  myfs::mainloop_nfs] lookup(81, "New folder")
    2024-05-07T17:23:19.710443Z DEBUG nfsserve::mount_handlers: 556337 --> mountres3_ok { fhandle: [131, 175, 247, 79, 143, 1, 0, 0, 85, 0, 0, 0, 0, 0, 0, 0], auth_flavors: [0, 1] }
    2024-05-07T17:23:19.710610Z TRACE nfsserve::rpcwire: Writing fragment length:60
    2024-05-07T17:23:19.710807Z DEBUG nfsserve::tcp: Message handling closed : Os { code: 10054, kind: ConnectionReset, message: "An existing connection was forcibly closed by the remote host." }
    2024-05-07T17:23:19.711085Z DEBUG nfsserve::tcp: Message loop broken due to early eof
     
    Stack backtrace:
       0: std::backtrace_rs::backtrace::dbghelp::trace
                 at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/../../backtrace/src/backtrace/dbghelp.rs:131:5
       ...
    2024-05-07T17:23:19.711790Z TRACE nfsserve::rpcwire: Reading fragment length:88, last:true
    2024-05-07T17:23:19.711921Z TRACE nfsserve::rpcwire: Finishing Reading fragment length:88, last:true
    2024-05-07T17:23:19.712083Z DEBUG nfsserve::nfs_handlers: nfsproc3_getattr(556338,nfs_fh3 { data: [131, 175, 247, 79, 143, 1, 0, 0, 85, 0, 0, 0, 0, 0, 0, 0] })
    [2024-05-07T17:23:19Z INFO  myfs::mainloop_nfs] getattr(85)
    2024-05-07T17:23:19.712371Z DEBUG nfsserve::nfs_handlers:  556338 --> fattr3 { ftype: NF3DIR, mode: 511, nlink: 1, uid: 507, gid: 507, size: 0, used: 0, rdev: specdata3 { specdata1: 0, specdata2: 0 }, fsid: 0, fileid: 1, atime: nfstime3 { seconds: 0, nseconds: 0 }, mtime: nfstime3 { seconds: 0, nseconds: 0 }, ctime: nfstime3 { seconds: 0, nseconds: 0 } }
    2024-05-07T17:23:19.712566Z TRACE nfsserve::rpcwire: Writing fragment length:112
    2024-05-07T17:23:19.712874Z  INFO nfsserve::tcp: Accepting connection from 127.0.0.1:956
    2024-05-07T17:23:19.713046Z DEBUG nfsserve::tcp: Accepting socket PollEvented { io: Some(TcpStream { addr: 127.0.0.1:111, peer: 127.0.0.1:956, socket: 728 }) } RPCContext { local_port: 111, client_addr: "127.0.0.1:956", auth: auth_unix { stamp: 0, machinename: [], uid: 0, gid: 0, gids: [] } }
    2024-05-07T17:23:19.713386Z TRACE nfsserve::rpcwire: Reading fragment length:100, last:true
    2024-05-07T17:23:19.713523Z TRACE nfsserve::rpcwire: Finishing Reading fragment length:100, last:true
    2024-05-07T17:23:19.713675Z DEBUG nfsserve::mount_handlers: mountproc3_umnt(556339,"//New folder/New folder")
    2024-05-07T17:23:19.713825Z TRACE nfsserve::rpcwire: Writing fragment length:28
    2024-05-07T17:23:19.714072Z DEBUG nfsserve::tcp: Message handling closed : Os { code: 10054, kind: ConnectionReset, message: "An existing connection was forcibly closed by the remote host." }
    2024-05-07T17:23:19.714383Z DEBUG nfsserve::tcp: Message loop broken due to early eof
     
    Stack backtrace:
       ...
    2024-05-07T17:23:19.715205Z TRACE nfsserve::rpcwire: Reading fragment length:156, last:true
    2024-05-07T17:23:19.715351Z TRACE nfsserve::rpcwire: Finishing Reading fragment length:156, last:true
    2024-05-07T17:23:19.715498Z DEBUG nfsserve::nfs_handlers: nfsproc3_create(556340, diropargs3 { dir: nfs_fh3 { data: [131, 175, 247, 79, 143, 1, 0, 0, 85, 0, 0, 0, 0, 0, 0, 0] }, name: "ab.txt" }, UNCHECKED)
    [2024-05-07T17:23:19Z INFO  myfs::mainloop_nfs] getattr(85)
    2024-05-07T17:23:19.715793Z DEBUG nfsserve::nfs_handlers: create unchecked sattr3 { mode: mode(493), uid: Void, gid: Void, size: size(0), atime: SET_TO_CLIENT_TIME(nfstime3 { seconds: 1715102599, nseconds: 715028000 }), mtime: SET_TO_CLIENT_TIME(nfstime3 { seconds: 1715102599, nseconds: 715028000 }) }
    [2024-05-07T17:23:19Z INFO  myfs::mainloop_nfs] create(85, "ab.txt", sattr3 { mode: mode(493), uid: Void, gid: Void, size: size(0), atime: SET_TO_CLIENT_TIME(nfstime3 { seconds: 1715102599, nseconds: 715028000 }), mtime: SET_TO_CLIENT_TIME(nfstime3 { seconds: 1715102599, nseconds: 715028000 }) })
    [2024-05-07T17:23:19Z INFO  myfs::mainloop_nfs] getattr(85)
    2024-05-07T17:23:19.722299Z DEBUG nfsserve::nfs_handlers: create success --> 87, attributes(fattr3 { ftype: NF3REG, mode: 511, nlink: 1, uid: 507, gid: 507, size: 0, used: 0, rdev: specdata3 { specdata1: 0, specdata2: 0 }, fsid: 0, fileid: 2, atime: nfstime3 { seconds: 0, nseconds: 0 }, mtime: nfstime3 { seconds: 0, nseconds: 0 }, ctime: nfstime3 { seconds: 0, nseconds: 0 } })
    2024-05-07T17:23:19.722513Z TRACE nfsserve::rpcwire: Writing fragment length:256
    2024-05-07T17:23:19.722902Z TRACE nfsserve::rpcwire: Reading fragment length:124, last:true
    2024-05-07T17:23:19.723091Z TRACE nfsserve::rpcwire: Finishing Reading fragment length:124, last:true
    2024-05-07T17:23:19.723235Z DEBUG nfsserve::nfs_handlers: nfsproc3_setattr(556341,SETATTR3args { object: nfs_fh3 { data: [131, 175, 247, 79, 143, 1, 0, 0, 87, 0, 0, 0, 0, 0, 0, 0] }, new_attribute: sattr3 { mode: Void, uid: Void, gid: Void, size: Void, atime: DONT_CHANGE, mtime: SET_TO_CLIENT_TIME(nfstime3 { seconds: 0, nseconds: 0 }) }, guard: Void })
    [2024-05-07T17:23:19Z INFO  myfs::mainloop_nfs] getattr(87)
    [2024-05-07T17:23:19Z INFO  myfs::mainloop_nfs] setattr(87, sattr3 { mode: Void, uid: Void, gid: Void, size: Void, atime: DONT_CHANGE, mtime: SET_TO_CLIENT_TIME(nfstime3 { seconds: 0, nseconds: 0 }) })
    2024-05-07T17:23:19.723701Z DEBUG nfsserve::nfs_handlers:  setattr success 556341 --> fattr3 { ftype: NF3REG, mode: 511, nlink: 1, uid: 507, gid: 507, size: 0, used: 0, rdev: specdata3 { specdata1: 0, specdata2: 0 }, fsid: 0, fileid: 2, atime: nfstime3 { seconds: 0, nseconds: 0 }, mtime: nfstime3 { seconds: 0, nseconds: 0 }, ctime: nfstime3 { seconds: 0, nseconds: 0 } }
    2024-05-07T17:23:19.723897Z TRACE nfsserve::rpcwire: Writing fragment length:144
    2024-05-07T17:23:19.724417Z TRACE nfsserve::rpcwire: Reading fragment length:92, last:true
    2024-05-07T17:23:19.724565Z TRACE nfsserve::rpcwire: Finishing Reading fragment length:92, last:true
    2024-05-07T17:23:19.724702Z DEBUG nfsserve::nfs_handlers: nfsproc3_access(556342,nfs_fh3 { data: [131, 175, 247, 79, 143, 1, 0, 0, 85, 0, 0, 0, 0, 0, 0, 0] },16)
    [2024-05-07T17:23:19Z INFO  myfs::mainloop_nfs] getattr(85)
    2024-05-07T17:23:19.724990Z DEBUG nfsserve::nfs_handlers:  556342 ---> 16
    2024-05-07T17:23:19.725149Z TRACE nfsserve::rpcwire: Writing fragment length:120
    2024-05-07T17:23:19.725378Z TRACE nfsserve::rpcwire: Reading fragment length:116, last:true
    2024-05-07T17:23:19.725535Z TRACE nfsserve::rpcwire: Finishing Reading fragment length:116, last:true
    2024-05-07T17:23:19.725679Z DEBUG nfsserve::nfs_handlers: nfsproc3_remove(556343, diropargs3 { dir: nfs_fh3 { data: [131, 175, 247, 79, 143, 1, 0, 0, 85, 0, 0, 0, 0, 0, 0, 0] }, name: "New Text Document.txt" })
    [2024-05-07T17:23:19Z INFO  myfs::mainloop_nfs] getattr(85)
    [2024-05-07T17:23:19Z INFO  myfs::mainloop_nfs] remove(85, "New Text Document.txt")
    [2024-05-07T17:23:19Z INFO  myfs::mainloop_nfs] lookup(85, "New Text Document.txt")
    [2024-05-07T17:23:19Z INFO  myfs::mainloop_nfs] getattr(85)
    2024-05-07T17:23:19.726400Z DEBUG nfsserve::nfs_handlers: remove success
    2024-05-07T17:23:19.726548Z TRACE nfsserve::rpcwire: Writing fragment length:144
```

Meanwhile log trace for the failed `Rename-Item` for folder or rename through windows explorer just stop at the second stack trace and there's no more print outs. So I guess the rpc request didn't even reach handle_nfs() part for delegation?

And interestingly I found multiple post on web mentioning that they see the same "invalid device" error when try to rename a file / folder through explorer : https://learn.microsoft.com/en-us/archive/msdn-technet-forums/1047f8fa-20fd-4c0c-b351-af5a7f778b13, and enable `showmount` for NFS server might workaround the problem. Not sure if this showmount should be enabled through our nfsserve code? I tried to enable it on my Windows OS directly but seems after I install nfs server through server manager the port :111 will be occupied and then I cannot mount the nfs server anymore through our lib anymore...

Besides the main blocker for me is I would like to have `Rename-Item` working for folder as well but it always give me access denied and the log trace is same as the explorer rename that didn't work...

PS: some other reference that it's mentions as bug for Windows 8 and Windows server 2012 and there's patch for it, but I am using Windows server 2022 so not sure why I still see this issue " https://support.microsoft.com/en-us/topic/-invalid-device-error-when-you-try-to-rename-a-file-on-a-network-file-system-client-that-is-running-windows-8-windows-8-1-windows-server-2012-or-windows-server-2012-r2-ba78e302-de3a-e528-8043-e2ad9fd689c0

Let me now if this helps, thanks!

**Comment by ylow** [comment]
Hmm. That's very interesting... I can't tell from the bug reports what is problem, or what protocol / command it is using. I need to lookup what does "showmount" actually do at the protocol level... Will do some googling... let me know if you find anything.

**Comment by freshmanken** [comment]
Thanks @ylow, do you know the folder's attributes (right click - properties - general - attributes) is always marked as Readonly even if I set the NFS attribute to 777? Tried to remove that mark but it never take into effect...Is there anyway we can set this property up through the NFS server? Suspect this might be the culprit of why `Rename-Item` for file is ok but `Rename-Item` for folder is not

**Comment by rrauch** [comment]
After hours of staring at Wireshark logs, I have been able to find the root cause of this issue:

This problem is caused by the NFS export path being set to `/` and the Windows client not liking it very much. My guess here is that for certain operations (of which apparently renaming is one of them), the Windows client first checks if the path belongs to the same export. When the path is `/`, this check does not work correctly and fails. This is clearly an issue of the Windows client and not `nfsserve` directly.

To fix this, the export name has to be set to `something`. I have submitted a PR here that allows setting a custom export name:

https://github.com/xetdata/nfsserve/pull/32

This seems to fix it.
